packages/pigeon/lib/cpp_generator.dart

```diff
@@ -0,0 +1,683 @@
+String _getSafeArgumentName(int count, NamedType argument) =>
+    _getArgumentName(count, argument) + 'Arg';
+
+void _writeFlutterApiHeader(Indent indent, Api api) {
+  assert(api.location == ApiLocation.flutter);
+  indent.writeln(
+      '/** Generated class from Pigeon that represents Flutter messages that can be called from C++.*/');
+  indent.write('class ${api.name} ');
+  indent.scoped('{', '};', () {
+    indent.scoped('private:', '', () {
+      indent.writeln('flutter::BinaryMessenger* binaryMessenger;');
+    });
+    indent.scoped('public:', '', () {
+      indent
+          .write('${api.name}(flutter::BinaryMessenger* argBinaryMessenger);');
+      // final String codecName = _getCodecName(api);
+      indent.writeln('');
+      indent.format('''
+  static const flutter::StandardMessageCodec& GetCodec();
+  ''');
+      for (final Method func in api.methods) {
+        final String returnType = func.returnType.isVoid
+            ? 'void'
+            : _cppTypeForDartType(func.returnType);
+        if (func.arguments.isEmpty) {
+          indent.write(
+              'void ${func.name}(std::function<void($returnType)>&& callback);');
+        } else {
+          final Iterable<String> argTypes =
+              func.arguments.map((NamedType e) => _cppTypeForDartType(e.type));
+          final Iterable<String> argNames =
+              indexMap(func.arguments, _getSafeArgumentName);
+          final String argsSignature =
+              map2(argTypes, argNames, (String x, String y) => '$x $y')
+                  .join(', ');
+          indent.write(
+              'void ${func.name}($argsSignature, std::function<void($returnType)>&& callback);');
+        }
+      }
+    });
+  });
+}
+
+void _writeFlutterApiSource(Indent indent, Api api) {
+  assert(api.location == ApiLocation.flutter);
+  indent.writeln(
+      '/** Generated class from Pigeon that represents Flutter messages that can be called from C++.*/');
+  indent.write(
+      '${api.name}::${api.name}(flutter::BinaryMessenger* argBinaryMessenger)');
+  indent.scoped('{', '}', () {
+    indent.writeln('this->binaryMessenger = argBinaryMessenger;');
+  });
+  final String codecName = _getCodecName(api);
+  indent.format('''
+const flutter::StandardMessageCodec& ${api.name}::GetCodec() {
+\treturn flutter::StandardMessageCodec::GetInstance(&$codecName::GetInstance());
+}
+''');
+  for (final Method func in api.methods) {
+    final String channelName = makeChannelName(api, func);
+    final String returnType =
+        func.returnType.isVoid ? 'void' : _cppTypeForDartType(func.returnType);
+    String sendArgument;
+    if (func.arguments.isEmpty) {
+      indent.write(
+          'void ${api.name}::${func.name}(std::function<void($returnType)>&& callback) ');
+      sendArgument = 'flutter::EncodableValue()';
+    } else {
+      final Iterable<String> argTypes =
+          func.arguments.map((NamedType e) => _cppTypeForDartType(e.type));
```

Comment:
I've updated this to use const ref for non-built-in types. Should we use std::optional to allow nullables? Maybe this would require C++20 for references, so I'm unsure about the right approach here.

Reply:
Ok, I gave it a shot, but we need more e2e tests to properly test this.